scheduler/core/components/optimizer/greedymax.py

```diff
@@ -15,18 +15,45 @@
 from .base import BaseOptimizer
 
 import numpy as np
-import astropy.units as u
+# import astropy.units as u
 
 
 class GreedyMaxOptimizer(BaseOptimizer):
     """
     GreedyMax is an optimizer that schedules the visits for the rest of the night in a greedy fashion.
     """
 
-    def __init__(self, seed=42):
+    def __init__(self):
         self.groups = []
-        self.groupids = []
+        self.group_ids = []
+        # self.obs_groups = []     # remove if not used
+        self.obs_group_ids = []
         self.timelines = []
+        self.sites = []
+        self.min_visit_len = timedelta(minutes=30)
+
+    def setup(self, selection: Selection) -> GreedyMaxOptimizer:
+        """
+        Preparation for the optimizer e.g. create chromosomes, etc.
+        """
+        self.groups = list(selection.schedulable_groups.values())
+        self.group_ids = list(selection.schedulable_groups)
+        for gid, group_data in selection.schedulable_groups.items():
+            if group_data.group.is_observation_group():
+                self.obs_group_ids.append(group_data.group.unique_id())
+                # self.obs_groups.append(group_data.group)
+            elif group_data.group.is_scheduling_group():
+                for subgroup in group_data.group.children:
+                    if subgroup.is_observation_group():
+                        self.obs_group_ids.append(subgroup.unique_id())
+                        # self.obs_groups.append(subgroup)
```

Comment:
Unfortunately _process_group_data doesn't work, I get the error "AttributeError: 'AndGroup' object has no attribute 'group'".  That is why I had to iterate over the children. I haven't figured out how to get this to work with recursion.

Reply:
Hmmm... that seems unusual, like we're getting passed the wrong data type. I'll take a look at it tomorrow morning. I think `group_data` is actually an `AndGroup`, which would explain the problem. I'll take a look at it tomorrow morning.

Reply:
Oh. *headslaps* I see what I'm doing wrong here.

Instead of this, why don't we add a `Selection.obs_group_ids`?

That way, we don't have to do this calculation here, and it is easier done in the Selector as a one-liner, since we go from this information specifically to only having the top-level (non-root) groups there. That way you can just grab it from the `Selection` and store it if you need it.

Reply:
I'm looking and I see you need this to be a `List` instead of a `Set` or `FrozenSet` since you use an `index` call into it. I'm a little curious why you're taking indices into this instead of just using the IDs themselves. Let me look at this in the morning.